### PR TITLE
STM32L5 Architecture Support

### PR DESCRIPTION
**Pull request by michi-jung**
## Summary

This is my current state of work on getting NuttX to run on STM32L5. I am especially interested in the TrustZone support of ARMV8-M. The 'nsh' configuration of the stm32l562e-dk board runs fine as the non-secure companion to STMicroelectronics' port of TrustedFirmware-M.

Original pull-request was https://github.com/apache/incubator-nuttx/pull/2970, which I think I messed up by force-pushing. Sorry for the fuzz.

## Impact

All code is in new subdirectories and should not have any impact on current configurations.

## Testing

I am running a basic 'nsh' configuration fine directly on a Nucleo-L552ZE-Q board and as a companion to TrustedFirmware-M on STM32L562E-DK.

### CODE REVIEW

**Comment by michi-jung** [comment]
The remaining files with a BSD license header are:

hardware/stm32l562xx_dbgmcu.h
hardware/stm32l5_flash.h
hardware/stm32l5_gpio.h
hardware/stm32l5_memorymap.h
hardware/stm32l5_pinmap.h
hardware/stm32l5_pwr.h
hardware/stm32l5_spi.h
hardware/stm32l5_syscfg.h
hardware/stm32l5_tim.h
hardware/stm32l5_uart.h
stm32l5_allocateheap.c
stm32l5_flash.c
stm32l5_flash.h
stm32l5_gpio.c
stm32l5_gpio.h
stm32l5_irq.c
stm32l5_lse.c
stm32l5_lsi.c
stm32l5_rcc.c
stm32l5_rcc.h
stm32l5_serial.c
stm32l5_spi.c
stm32l5_spi.h
stm32l5_tim.c
stm32l5_tim.h
stm32l5_tim_lowerhalf.c
stm32l5_uart.h

Most of them can get cleared if we get ICLAs or an SGA from the Haltian team.

**Comment by michi-jung** [comment]
I think I am happy with the pull request now.  We can iron out the remaining BSD license headers once respective ICLAs are available. Could you please give me feedback on what remains to be done to get this upstream? Thanks!

**Comment by xiaoxiang781216** [comment]
@davids5 could you take a look? you are STM expert.

**Comment by davids5** [comment]
@michi-jung @xiaoxiang781216 - yes but it will take me a while to get to it.

**Comment by acassis** [comment]
@xiaoxiang781216 also we need to request @raiden00pl to take a look!

File: arch/arm/src/stm32l5/chip.h
Comment (acassis) [inline arch/arm/src/stm32l5/chip.h:54]: @michi-jung shouldn't the cache line be 32 bytes like in other STM32xx ?

Comment (michi-jung) [reply arch/arm/src/stm32l5/chip.h:54]: The Cortex-M33 used in STM32L5 does not include any D-Cache or I-Cache.  For other architectures that include e.g. a Cortex-M7 with a D-Cache those literals are important for cache aligned DMA operations.  Actually, both those defines are not used in the complete stm32l5 code. I am going to remove them. Thanks!

**Comment by acassis** [comment]
Hi @michi-jung except from Haltian licensing issue, is there something else missing to merge it? I'm considering to move ahead and later when Haltian submit the SGA we fix the files' license. @raiden00pl did you have the chance to review it?

**Comment by raiden00pl** [comment]
I looked through most of the files and LGTM

File: arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h
Comment (davids5) [inline arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h:71]: Formatting is not consistent

File: arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h
Comment (davids5) [inline arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h:442]: I blew this on the F7/H7 and it is coming back to bite us on EMI and overshoot.

Remove the drive strengths on all GPIO, all GPIO will need to have a _n - the board config should set this, using the _n |   GPIO_SPEED_xxMHz it needs.

File: arch/arm/src/stm32l5/hardware/stm32l562xx_rcc.h
Comment (davids5) [inline arch/arm/src/stm32l5/hardware/stm32l562xx_rcc.h:81]: formatting  is not consistent

File: arch/arm/src/stm32l5/hardware/stm32l5_exti.h
Comment (davids5) [inline arch/arm/src/stm32l5/hardware/stm32l5_exti.h:63]: " 
some files have (N+O)  some have (N + O)

File: arch/arm/src/stm32l5/stm32l5_lse.c
Comment (davids5) [inline arch/arm/src/stm32l5/stm32l5_lse.c:141]: This can be an infinite loop if the LSECLOCK_RUN_DRV_CAPABILITY is not set high enough.  See https://github.com/apache/incubator-nuttx/pull/2943 and https://github.com/apache/incubator-nuttx/pull/2967 https://github.com/apache/incubator-nuttx/pull/2969

**Review by davids5 — CHANGES_REQUESTED** [review]
@michi-jung - nice work!

Disclaimer: I did not review the datasheet to the code.

There are some formatting nits.

You may want to adopt the LSE changes. 

But most critical is the GPIO changes. I set a bad precedent, and you have an opportunity to fix it. The board should set all signal attributes (SPEED etc).

**Comment by michi-jung** [comment]
Hi @acassis, I will try to resolve @davids5's review comments tonight. Once that is done its fine with me to go ahead and merge it. Thanks!

File: arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h
Comment (michi-jung) [reply arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h:71]: You mean because of the blank after GPIO_AF9, right? Ok, will remove all blanks. Thanks.

File: arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h
Comment (michi-jung) [reply arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h:442]: Makes sense. I will remove all _SPEED_ constants. Thanks!

File: arch/arm/src/stm32l5/hardware/stm32l5_exti.h
Comment (michi-jung) [reply arch/arm/src/stm32l5/hardware/stm32l5_exti.h:63]: I grep'ed all the header files and they all seem to be (N+O).  Can you point me to an (N + O) case, please? Thanks.

File: arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h
Comment (davids5) [reply arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h:71]: yes

**Review by davids5 — CHANGES_REQUESTED** [review]

File: arch/arm/src/stm32l5/hardware/stm32l5_exti.h
Comment (michi-jung) [reply arch/arm/src/stm32l5/hardware/stm32l5_exti.h]: I think I addressed this with https://github.com/apache/incubator-nuttx/pull/2974/commits/66f6a22c40060bd66b0f5fe6053b1dbfb70bf326. Can you please have a look?

File: arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h
Comment (michi-jung) [reply arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h]: Please have a look at https://github.com/apache/incubator-nuttx/pull/2974/commits/70ffe11526af37f412e475063a7b690ed1e5fa3d

File: arch/arm/src/stm32l5/stm32l5_lse.c
Comment (michi-jung) [reply arch/arm/src/stm32l5/stm32l5_lse.c:153]: Please have a look at https://github.com/apache/incubator-nuttx/pull/2974/commits/8767cf85523812fc7e0cc6de97c1829065a9a98a

File: arch/arm/src/stm32l5/hardware/stm32l562xx_rcc.h
Comment (michi-jung) [reply arch/arm/src/stm32l5/hardware/stm32l562xx_rcc.h]: Please have a look at https://github.com/apache/incubator-nuttx/commit/66f6a22c40060bd66b0f5fe6053b1dbfb70bf326 (The same commit applies to your next review finding)

File: arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h
Comment (michi-jung) [reply arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h]: Please have a look at https://github.com/apache/incubator-nuttx/pull/2974/commits/233c38d9adce7a28767bb2edfabb772df0d0fe5b

**Comment by michi-jung** [comment]
The 'Build / Fetch-Source (pull_request)' step now shows a lot of coding-style problems, which are not in the files I modified but in other files that have meanwhile been touched in apache:master. Shall I re-base and force push this pull-request?

**Comment by xiaoxiang781216** [comment]
> The 'Check / check (pull_request)' step now shows a lot of coding-style problems, which are not in the files I modified but in other files that have meanwhile been touched in apache:master. Shall I re-base and force push this pull-request?

yes, please:
```
git fetch apache
git rebase apache/master
git push -f origin stm32l5-upstreaming
```

**Comment by davids5** [comment]
@michi-jung  - the rest of the CI fails are in this PR. One more fixup and force push should do it,

File: arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h
Comment (davids5) [reply arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h]: it left spaces

File: arch/arm/src/stm32l5/stm32l5_lse.c
Comment (davids5) [reply arch/arm/src/stm32l5/stm32l5_lse.c:153]: It is better than the dead wait. But will not ramp to get it started. - I do not feel strongly about ramping herein at this point.

File: arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h
Comment (michi-jung) [reply arch/arm/src/stm32l5/hardware/stm32l562xx_pinmap.h]: Spaces have been removed with separate commit https://github.com/apache/incubator-nuttx/commit/70759650d9cf4d499a75e22a8f0eb91a5c07bb0d

**Review by davids5 — APPROVED** [review]
@michi-jung  - looks good to go once CI finishes. 
Once this is merged, we need to do a PR on the https://github.com/apache/incubator-nuttx-testing/commits/master to add the board to the build FYI @xiaoxiang781216

**Comment by michi-jung** [comment]
Hi @acassis, I believe we can merge this now. I will provide the LSE drive ramp-up in a separate pull request. And the headers can be changed to ASF once the remaining license issues are cleared.

Thanks to @acassis, @davids5, @xiaoxiang781216 and @raiden00pl for your help!

**Pull request merged**
